azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -1599,30 +1599,27 @@ public void run() {
                         executor.getPort(), ConnectorParams.UPDATE_ACTION,
                         null, null, executionIds, updateTimes);
               } catch (final ExecutorManagerException e) {
-                logger.error(e);
+                logger.error("Failed to get update from executor " + executor.getHost(), e);
+                boolean sendUnresponsiveEmail = false;
                 for (final ExecutableFlow flow : entry.getValue()) {
                   final Pair<ExecutionReference, ExecutableFlow> pair =
                       ExecutorManager.this.runningFlows.get(flow.getExecutionId());
 
                   ExecutorManager.this.updaterStage =
-                      "Failed to get update. Doing some clean up for flow "
-                          + pair.getSecond().getExecutionId();
-
-                  if (pair != null) {
-                    final ExecutionReference ref = pair.getFirst();
-                    int numErrors = ref.getNumErrors();
-                    if (ref.getNumErrors() < this.numErrors) {
-                      ref.setNextCheckTime(System.currentTimeMillis()
-                          + this.errorThreshold);
-                      ref.setNumErrors(++numErrors);
-                    } else {
-                      logger.warn("Evicting execution " + flow.getExecutionId()
-                          + ". The executor is unresponsive.");
-                      // TODO should send out an unresponsive email here.
-                      finalizeFlows.add(pair.getSecond());
-                    }
+                      "Failed to get update for flow " + pair.getSecond().getExecutionId();
+
+                  final ExecutionReference ref = pair.getFirst();
+                  ref.setNextCheckTime(System.currentTimeMillis() + this.errorThreshold);
+                  ref.setNumErrors(ref.getNumErrors() + 1);
+                  if (ref.getNumErrors() % this.numErrorsBeforeUnresponsiveEmail == 0) {
+                    // if any of the executions has failed many enough updates, alert
+                    sendUnresponsiveEmail = true;
                   }
                 }
+                if (sendUnresponsiveEmail) {
+                  final Alerter mailAlerter = ExecutorManager.this.alerterHolder.get("email");
+                  mailAlerter.alertOnFailedUpdate(executor, entry.getValue(), e);
```

Comment:
Will users be flooded with alerting emails when the executor is unresponsive? UpdaterThread checks the executor status frequently and it might keep sending emails before we could take the action to remove the executor from DB.

Reply:
Yes, it would send the emails ~every 1 minute.

```
// When we have an http error, for that flow, we'll check every 10 secs, 6
// times (1 mins) before we send an email about unresponsive executor.
// private final int numErrorsBeforeUnresponsiveEmail = 6;
```

The idea is to send e-mail again if the problem remains, but this is clearly too often.

 How about sending these emails once per hour, ie. change to `numErrorsBeforeUnresponsiveEmail = 60`?